src/docs_pseudo_methods.cr

```diff
@@ -0,0 +1,197 @@
+#
+# ```
+# # On a 64 bits machine
+# sizeof(Pointer(Int32)) # => 8
+# sizeof(String)         # => 8
+# ```
+#
+# This is because a `Reference`'s memory is allocated on the heap and a pointer
+# to it is passed around. The effective size of a class can be determined using
+# `instance_sizeof`.
+def __crystal_pseudo_sizeof(type : Class) : Int32
+end
+
+# Returns the instance size of the given class as number of bytes.
+#
+# *type* must be a constant, it cannot be evaluated at runtime.
+#
+# ```
+# instance_sizeof(String)    # => 16
+# instance_sizeof(Exception) # => 48
+# ```
+#
+# See `sizeof` for determining the size of value types.
+def __crystal_pseudo_instance_sizeof(type : Class) : Int32
+end
+
+# Returns a `Pointer` to the contents of a variable.
+#
+# *variable* must be a variable (local, instance, class or library).
+#
+# ```
+# a = 1
+# ptr = pointerof(a)
+# ptr.value = 2
+#
+# a # => 2
+# ```
+def __crystal_pseudo_pointerof(variable : T) : Pointer(T) forall T
+end
+
+# Returns the byte offset of an instance variable in a struct or class type.
+#
+# *type* must be a constant, it cannot be evaluated at runtime.
+# *variable*  must be the name of an instance variable of *type*, prefixed
+# by `@`.
+# ```
+# offsetof(String, @bytesize)   # => 4
+# offsetof(Exception, @message) # => 8
+# offsetof(Time, @location)     # => 16
+# ```
+def __crystal_pseudo_offsetof(type : Class, variable) : Int32
+end
+
+class Object
+  # Returns the boolean negation of `self`.
+  #
+  # ```
+  # !true  # => false
+  # !false # => true
+  # !nil   # => true
+  # !1     # => false
+  # !"foo" # => false
+  # ```
+  #
+  # This method is a unary operator and usually written in prefix notation
+  # (`!foo`) but it can also be written as a regular method call (`foo.!`).
+  def __crystal_pseudo_! : Bool
+  end
+
+  # Returns `true` if `self` inherits or includes *type*.
+  # *type* must be a constant, it cannot be evaluated at runtime.
+  #
+  # ```
+  # a = 1
+  # a.class                 # => Int32
+  # a.is_a?(Int32)          # => true
+  # a.is_a?(String)         # => false
+  # a.is_a?(Number)         # => true
+  # a.is_a?(Int32 | String) # => true
+  # ```
+  def __crystal_pseudo_is_a?(type : Class) : Bool
+  end
+
+  # Returns `true` if `self` is `Nil`.
+  #
+  # ```
+  # 1.nil?   # => false
+  # nil.nil? # => true
+  # ```
+  #
+  # This method is equivalent to `is_a?(Nil)`.
+  def __crystal_pseudo_nil? : Bool
+  end
+
+  # Returns `self`.
+  #
+  # The type of this expression is restricted to *type* by the compiler.
+  # *type* must be a constant, it cannot be evaluated at runtime.
+  #
+  # If *type* is not a valid restriction for the expression type, it
+  # is a compile-time error.
+  # If *type*  is a valid restriction for the expression, but `self` can't
+  # be restricted to *type*, it raises at runtime.
+  # *type* may be a wider restriction than the expression type, the resulting
```

Comment:
Not entirely true (or there is actually a bug :-) )

```
a : Int32
typeof(a.as(Int32 | String)) # => Int32 | String (yet the value is Int32)

b : Int32 | String
typeof(b.as(Int32 | Float64)) # => Int32
typeof(b.as(Int32 | Float64).as(Int32 | Float64)) # => Int32 | Float64
```

https://play.crystal-lang.org/#/r/7u51

Reply:
You can use `as` to extend the type of an expression from a single type to a union type. The `b` case is a bit strange but I guess it also makes sense (check that it's an `Int32` and then extends it to `Int32 | Float64`, raise otherwise).

Reply:
What is weird is that 

* `expr.as(T) : T` does not hold
* `expr.as?(T) : T?` does not hold
* `expr.as(T) == expr.as(T).as(T)` does not hold

of course if `expr` was a value that is not compatible with `T` then raise / nil is expected. But I don't see why the above rules cannot hold


It is definitely something beyond the documentation. But I picked from the phrases and the examples 🕵

Reply:
Oh, yes, `b.as(Int32 | Float64)` should give `Int32 | Float64`. It's a bug. Sorry, I missed that.

Actually, I don't think casting between union types like that should be allowed, it's strange.